File: 01pathtoone/solution.py
import sys
from Queue import Queue

n = int(sys.stdin.readline())

queue = Queue()

queue.put((n , 0))

while not queue.empty():
    (x, count) = queue.get()
    queue.task_done()
    if x is 1:
        print count
        break

    if x % 2 is 0:
        next_x = x // 2
        queue.put((x // 2, count + 1))
    else:
        queue.put((x + 1, count + 1))
        queue.put((x - 1, count + 1))

File: 01pathtoone/solution.cpp
#include<iostream>
#include<queue>


int main() {
    int n;

    std::cin >> n;

    std::queue<int> q;

    q.push(n);
    q.push(0);

    while(!q.empty()) {
        int current = q.front();
        q.pop();
        int count = q.front();
        q.pop();

        if(current == 1) {
            std::cout << count << '\n';
            break;
        }

        if(current % 2 == 0) {
            q.push(current / 2);
            q.push(count + 1);
        } else {
            q.push(current + 1);
            q.push(count + 1);
            q.push(current - 1);
            q.push(count + 1);
        }
    }

    return 0;
}

